### PR TITLE
Guidance to avoid duplicate metrics in multiprocess

### LINKED ISSUE #647: multiprocess in uwsgi leads generate_latest to sometimes contain duplicates
Question: can the `generate_latest` method detect and fix duplicates?

The code for my app is pretty simple. At the top of `middleware.py` I do:

```
# Metrics
# See https://github.com/amitsaha/python-prometheus-demo/blob/master/flask_app_prometheus_multiprocessing/src/helpers/middleware.py
registry = CollectorRegistry()
multiprocess.MultiProcessCollector(registry)
REQUEST_COUNT = Counter(
    "indexapi_request_count",
    "help: App Request Count",
    ["method", "endpoint"],
    registry=registry,
)


def setup_metrics(app):
    # https://github.com/prometheus/client_python

    @app.route("/metrics")
    def metrics():
        data = generate_latest(registry)
        print(data)
        return Response(data, mimetype=CONTENT_TYPE_LATEST)

```

somewhere in `middleware.py` during `_call` I do: 

```
REQUEST_COUNT.labels(request.method, path).inc()
```
in dockerfile I have:
```
RUN mkdir /tmp/prom_data
ENV prometheus_multiproc_dir /tmp/prom_data
```

In my main `app.py` I have:

```
app = Flask(__name__)
app.wsgi_app = middleware.Middleware(app.wsgi_app)
CORS(app)

logger.debug("Starting metrics")
middleware.setup_metrics(app)
```

I run UWSGI with:

```
CMD uwsgi --http 0.0.0.0:8080 --wsgi-file indexapi/app.py --callable app --buffer-size 30000  --processes 4 --threads 4
```

It seems to work fine, as in no matter what thread processes the request, the metric gets updated once. However, sometimes the text contains duplicate output:

```
 ~/D/python-indexapi   prom    curl localhost:4444/metrics                                                                            Mon May  3 12:36:28 2021
# HELP indexapi_request_count_total Multiprocess metric
# TYPE indexapi_request_count_total counter
indexapi_request_count_total{endpoint="/index",method="GET"} 1.0
# HELP indexapi_request_count_total help: App Request Count
# TYPE indexapi_request_count_total counter

 ~/D/python-indexapi   prom    curl localhost:4444/metrics                                                                            Mon May  3 12:36:29 2021
# HELP indexapi_request_count_total Multiprocess metric
# TYPE indexapi_request_count_total counter
indexapi_request_count_total{endpoint="/index",method="GET"} 1.0
# HELP indexapi_request_count_total help: App Request Count
# TYPE indexapi_request_count_total counter
indexapi_request_count_total{endpoint="/index",method="GET"} 1.0
# HELP indexapi_request_count_created help: App Request Count
# TYPE indexapi_request_count_created gauge
indexapi_request_count_created{endpoint="/index",method="GET"} 1.6200591934205577e+09

 ~/D/python-indexapi   prom    curl localhost:4444/metrics                                                                            Mon May  3 12:36:29 2021
# HELP indexapi_request_count_total Multiprocess metric
# TYPE indexapi_request_count_total counter
indexapi_request_count_total{endpoint="/index",method="GET"} 1.0
# HELP indexapi_request_count_total help: App Request Count
# TYPE indexapi_request_count_total counter
```

This is very perplexing to me. We can see 

```
# TYPE indexapi_request_count_total counter
indexapi_request_count_total{endpoint="/index",method="GET"} 1.0
# HELP indexapi_request_count_total help: App Request Count

# TYPE indexapi_request_count_total counter
indexapi_request_count_total{endpoint="/index",method="GET"} 1.0
# HELP indexapi_request_count_created help: App Request Count
```

Also, when the duplicate is detected, this `gauge` appears out of nowwhere.

### PR DESCRIPTION
**Pull request by csmarchbanks**
Closes #647

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -504,6 +504,8 @@ To handle this the client library can be put in multiprocess mode.
 This comes with a number of limitations:
 
 - Registries can not be used as normal, all instantiated metrics are exported
+  - Registering metrics to a registry later used by a `MultiProcessCollector`
```
Comment (tommyjcarpenter) [inline README.md:507]: this is under a section called `(Gunicorn)` however this seems more generic and not tied to Gunicorn, right? I hit this issue with UWSGI.

File: README.md [hunk 2]
```diff
@@ -522,8 +524,12 @@ and not directly from Python (otherwise it may not propagate to child processes)
...
 **2. Metrics collector**:
 
-The application must initialize a new `CollectorRegistry`,
-and store the multi-process collector inside.
+The application must initialize a new `CollectorRegistry`, and store the
+multi-process collector inside. It is a best practice to create this registry
+inside the context of a request to avoid metrics registering themselves to a
+collector used by a `MultiProcessCollector`. If a registry with metrics
+registered is used by a `MultiProcessCollector` duplicate metrics may be
+exported, one for multiprocess, and one for the process serving the request.
```
Comment (tommyjcarpenter) [inline README.md:532]: i would add an example Counter metric to the below snippet, that does not use registry=

File: README.md (hunk 1 above)
Comment (tommyjcarpenter) [reply README.md:507]: this also begs the question of, what is the case someone does want to register a metric to a registry?

Comment (csmarchbanks) [reply README.md:507]: The Gunicorn specification is a good point, perhaps `(E.g. Gunicorn)` instead to make it clear that is an example?

There are lots of valid use cases to register a metric when in non-multiprocess mode, it just doesn't work for multi-process right now. For example, when building a custom exporter you might have one registry for internal metrics, and a second registry for the metrics coming from something being probed.

File: README.md (hunk 2 above)
Comment (csmarchbanks) [reply README.md:532]: :+1: I added one.

File: README.md (hunk 1 above)
Comment (tommyjcarpenter) [reply README.md:507]: roger that

File: README.md [hunk 3]
```diff
@@ -504,30 +504,38 @@ To handle this the client library can be put in multiprocess mode.
...
+    may cause duplicate metrics to be exported
 - Custom collectors do not work (e.g. cpu and memory metrics)
 - Info and Enum metrics do not work
 - The pushgateway cannot be used
 - Gauges cannot use the `pid` label
 
 There's several steps to getting this working:
 
-**1. Gunicorn deployment**:
+**1. Deployment**:
```
Comment (tommyjcarpenter) [inline README.md:516]: :+1:

File: README.md [hunk 4]
```diff
@@ -504,30 +504,38 @@ To handle this the client library can be put in multiprocess mode.
...
+collector used by a `MultiProcessCollector`. If a registry with metrics
+registered is used by a `MultiProcessCollector` duplicate metrics may be
+exported, one for multiprocess, and one for the process serving the request.
 
 ```python
 from prometheus_client import multiprocess
-from prometheus_client import generate_latest, CollectorRegistry, CONTENT_TYPE_LATEST
+from prometheus_client import generate_latest, CollectorRegistry, CONTENT_TYPE_LATEST, Counter
+
+MY_COUNTER = Counter('my_counter', 'Description of my counter')
```
Comment (tommyjcarpenter) [inline README.md:538]: :+1:

**Review by tommyjcarpenter — APPROVED** [review]

**Pull request merged**